mobile/src/main/java/rs/readahead/washington/mobile/util/jobs/TellaUploadJob.java

```diff
@@ -0,0 +1,158 @@
+        if (keyBundle == null) { // CacheWord is unavailable
+            return exit(Job.Result.RESCHEDULE);
+        }
+
+        byte[] key = keyBundle.getKey();
+        if (key == null) { // key disappeared
+            return exit(Job.Result.RESCHEDULE);
+        }
+
+        dataSource = DataSource.getInstance(getContext(), key);
+        List<MediaFile> mediaFiles = dataSource.getUploadMediaFiles(ITellaUploadsRepository.UploadStatus.SCHEDULED).blockingGet();
+
+        if (mediaFiles.size() == 0) {
+            return exit(Job.Result.SUCCESS);
+        }
+
+        TellaUploadServer server = dataSource.getTellaUploadServer(Preferences.getAutoUploadServerId()).blockingGet();
+        if (server == TellaUploadServer.NONE) {
+            return exit(Result.FAILURE);
+        }
+        final TUSClient tusClient = new TUSClient(getContext(), server.getUrl(), server.getUsername(), server.getPassword());
+
+        for (MediaFile file : mediaFiles) {
+            fileMap.put(String.valueOf(file.getFileName()), file);
+        }
+
+        Flowable.fromIterable(mediaFiles)
+                .flatMap(tusClient::upload)
+                .blockingSubscribe(this::updateProgress, throwable -> {
+                    if (throwable instanceof NoConnectivityException) {
+                        exitResult = Result.RESCHEDULE;
+                        return;
+                    }
+                    Timber.d(throwable);
+                    Crashlytics.logException(throwable);
+                });
+
+        if (exitResult != null) {
+            exit(exitResult);
+        }
+
+        return exit(Job.Result.SUCCESS);
+    }
+
+    @Override
+    protected void onReschedule(int newJobId) {
+    }
+
+    private boolean enter() {
+        synchronized (TellaUploadJob.class) {
+            boolean current = running;
+            running = true;
+            return !current;
+        }
+    }
+
+    private Job.Result exit(Job.Result result) {
+        synchronized (TellaUploadJob.class) {
+            running = false;
+            return result;
+        }
+    }
+
+    public static void scheduleJob() {
+        new JobRequest.Builder(TellaUploadJob.TAG)
+                .setExecutionWindow(1_000L, 10_000L) // start between 1-10sec from now
+                .setBackoffCriteria(10_000L, JobRequest.BackoffPolicy.LINEAR)
+                .setRequiredNetworkType(JobRequest.NetworkType.CONNECTED)
+                .setRequirementsEnforced(true)
+                //.setPersisted(true) // android.Manifest.permission.RECEIVE_BOOT_COMPLETED
+                .setUpdateCurrent(true) // also, jobs will sync them self while running
+                .build()
+                .schedule();
+    }
+
+    private void updateProgress(UploadProgressInfo progressInfo) {
+
+        if (fileMap.get(progressInfo.name) == null) return;
+
+        MediaFile mediaFile = fileMap.get(progressInfo.name);
+        assert mediaFile != null;
```

Comment:
`mediaFile` can not be `null` here. Also, makes much more sense to use `RawFile` here and to use `id` instead of file name  as hash key. We should not assume here that file names are unique.

Reply:
I agree with all the objections, but I made compromises. I used the existing function and handler to delete a media file instead of creating new ones just for this purpose. Also, tusClient gives just a file name as a file identification (indeed it is unique in this case), and asserting that the media file is not null is just to make a linter happy.

Reply:
Yes, but if you put that `if (fileMap.get(progressInfo.name) == null) return;`, which is already checking is `mediaFile == null`, below the `MediaFile mediaFile = fileMap.get(progressInfo.name);` line - linter will be happy :). That is why I said `mediaFile` can not be null here and assert is not needed.

Reply:
For the file name uniqueness not in place here - `UploadProgressInfo` can be extended (new constructor accepting RawFile and taking name and id) to hold RawFile's database id also, which is unique and can be used here and avoid future bug.